LibWeb: Fix CSS "filter" clamping and resolution

BenWiederhake opened the pull request:
This was originally implemented in #1963. In particular, clamping was completely turned off:
6c642d168d54ced889546a88ffc0662f6d8564d2

However, *some* filter functions *do* need to be clamped, for example grayscale:
https://www.w3.org/TR/filter-effects-1/#funcdef-filter-grayscale

This made us fail at least the following test, which now passes (with this PR): http://wpt.live/css/filter-effects/filter-grayscale-005.html

Also tested against https://tweakers.net/ , like in #1963.

Ping @gmta, what do you think? You didn't include a test, so I'm not entirely sure whether this preserves all the things that you paid attention to. However, here's a screenshot of https://tweakers.net/ , just to demo that it didn't break anything obvious:

![Bildschirmfoto_2024-10-28_01-21-17](https://github.com/user-attachments/assets/b635c1b6-b3d5-49e2-be52-1cd3ebbca229)
![Bildschirmfoto_2024-10-28_01-24-41](https://github.com/user-attachments/assets/34d76319-bc13-4d6f-9ae4-0637f40498c8)

Found while investigating a very related, but ultimately different bug.

gmta commented:
In my PR, I only changed the interpretation of the filter property (i.e. clamping the result), not the actual computed value. This seems fine!

gmta commented on Tests/LibWeb/Text/input/css/filter-grayscale.html:
```diff
@@ -0,0 +1,24 @@
+        background-color: blue;
+        height: 30px;
+        width: 30px;
+        filter: grayscale(200%) brightness(200%) contrast(200%) invert(200%) opacity(200%) saturate(200%) sepia(200%);
```
Most of these filter functions also have lower boundaries ("cannot be negative"), maybe add a #low as well?

BenWiederhake replied:
It seemed like this was handled elsewhere, but now I'm not sure anymore. I'll check later.

gmta replied:
Checked it with this:
```js
let elm = document.createElement('a');
elm.style.filter = 'grayscale(-50%)';
console.log(elm.style.filter);
```

Gives me '' on Chrome and Firefox, but 'grayscale(-50%)' in Ladybird.

BenWiederhake replied:
Firefox displays this as an invalid value in the developer console, and therefore ignores it entirely. This means we probably shouldn't clamp it, either. Maybe the right thing is to reject it while parsing?

gmta replied:
I don't see that happening there, what version of Firefox are you testing this in? It looks like `grayscale(0%)` and below (negative values) effectively get ignored.

gmta replied:
Right, I was looking at the console output, not the style tab of an element. Don't think we should look at Firefox though, it was just a comparison. I'd rather we implement the specs (which calls negative values invalid).

BenWiederhake replied:
Absolutely! I just don't really know where to start with the spec, and since the common behavior seems to be to clamp values during parsing and drop invalid values, I suggest we do the same.

gmta replied:
I guess just copy and paste some of the constraints outlined here as spec comments and implement them?
https://drafts.fxtf.org/filter-effects/#supported-filter-functions

AtkinsSJ replied:
Yeah. For example, `grayscale()` says:

> Values of amount over 100% are allowed but UAs must clamp the values to 1.

So, you'd leave them as they are when parsing, and clamp them later. And...

> Negative values are not allowed.

So, you'd reject them while parsing (by returning null).

These constraints also apply to the result of `calc()` values, but it looks like our code doesn't allow them in filters. Once it does, `calc()` can't be bounds-checked when parsed, so it's always allowed (as long as it's the correct type) and then you'd need to clamp its value when it's used. (So if the `calc()` resolves to a negative number, you'd treat it as 0.)

gmta commented on Tests/LibWeb/Text/input/css/filter-grayscale.html:
```diff
@@ -0,0 +1,24 @@
+        background-color: blue;
+        height: 30px;
+        width: 30px;
+        filter: grayscale() brightness() contrast() invert() opacity() saturate() sepia();
```
Why not include functions like `hue-rotate` and `drop-shadow` as well?

BenWiederhake replied:
Because I only touched the filters implemented as "color" filters, which is exactly this subset.

gmta replied:
Fair enough! That's a FIXME then.

BenWiederhake commented:
Changes since last push:
- Rebased to current master
- Added spec links for the clamping logic
- Renamed "filter-grayscale.html" to "filter-color.html" to reflect that this is a test of only the basic color filters, not all filters in existence.

Note that implementing `calc()` is out-of-scope for this PR. (`filter: grayscale(calc(60%)) brightness(var(--myvariable));` works in Firefox and Chrome, so have fun replicating that behavior.)

tcl3 commented:
This has conflicts.

BenWiederhake commented:
Changes since last push:
- Rebased to current master to fix conflict, e.g. with 0de9818470d1b4c0b2aba3c88d7c87b95f5a0bf2
- Added a test with negative values
- Fixed the implementation of default stringification (e.g. `grayscale()` instead of `grayscale(1)`)
- Added a FIXME as requested; thanks @gmta! (Note that something about the existing blur is broken, so I don't want to touch it just yet.)

BenWiederhake commented:
Cannot work, since it crashes on `calc()` currently. However, that cannot be implemented, because `NumberPercentage` does not have a `::resolve` like its sibling types, due to:

```
/home/user/workspace/ladybird/Userland/Libraries/LibWeb/CSS/PercentageOr.h:101:40: error: ‘class Web::CSS::Number’ has no member named ‘percentage_of’
  101 |                 return reference_value.percentage_of(percentage);
      |                        ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/home/user/workspace/ladybird/Userland/Libraries/LibWeb/CSS/PercentageOr.h:104:45: error: ‘resolve_calculated’ is not a member of ‘Web::CSS::Number’
  104 |                 return T::resolve_calculated(calculated, layout_node, reference_value);
      |                        ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Now you might think "Just implement it, then!", but the spec specifically disallows this:

https://drafts.csswg.org/css-values-4/#example-10c2940e

> A <number-percentage> will never be added, as [<number>](https://drafts.csswg.org/css-values-4/#number-value) and [<percentage>](https://drafts.csswg.org/css-values-4/#percentage-value) can’t be combined in [calc()](https://drafts.csswg.org/css-values-4/#funcdef-calc).

So I think the only way forward is to disallow `calc()` in these simple "Color" filters.

gmta commented:
@BenWiederhake it makes sense that you can't end up with any of the other types in a calc, but surely we can `calc()` with `number-percentage`? Prohibiting calc does not happen in the other browsers either right?

BenWiederhake commented:
> Prohibiting calc does not happen in the other browsers either right?

Yes, it does:
![Bildschirmfoto_2024-11-07_14-59-53](https://github.com/user-attachments/assets/9169c635-ad55-4afc-94c2-6ef1e6847690)

However, trivial calcs (and probably vars) *are* supported, at least by Firefox:
![Bildschirmfoto_2024-11-07_15-01-51](https://github.com/user-attachments/assets/7ea23d52-0785-4ee3-a45b-26ef65e827d0)

I don't understand. Is there a section of the spec that allows this kind of unwrapping of trivial calcs?

Suggestion: We just parse calcs as invalid syntax. This fixes a WPT test, and improves the situation.

BenWiederhake commented:
I give up; this all seems too contradictory for me.

AtkinsSJ commented:
> I give up; this all seems too contradictory for me.

Welcome to CSS. 🙃

Sorry, I've not paid enough attention here.

Psychpsyo commented:
I think trivial calc() values might be fine in other browsers because I'm pretty sure they simplify those to the percentage itself. (which, of course, comes with a bunch of fun issues whenever the spec demands they be treated differently)
